### PR TITLE
Added in a blacklist feature

### PR DESCRIPTION
**Pull request by CactiChameleon9**
I managed to do it somehow. I have not been able to test the lazy images (I don't know how) but in theory the code following the if statement runs.

I deleted the zip because it needs updating and that is something I have no idea how to do :D

### CODE REVIEW

File: js/lazy_images.js [hunk 1]
```diff
@@ -17,6 +18,10 @@ function IsImageOk(img) {
   return true;
 }
 
+function arrayContains(needle, arrhaystack)
```
Comment (simonfrey) [inline js/lazy_images.js]: I think this can be substitued with the array.includes() method: 
https://www.w3schools.com/Jsref/jsref_includes_array.asp

File: js/lazy_images.js [hunk 2]
```diff
@@ -27,98 +32,100 @@ function setImgLazyLoad(result) {
...
-
-    }
-
-
-
-    browser.runtime.sendMessage({
-      lazyDone: true,
-      ourLazyLoad:   ourLazy
-    });
+	  if (!(arrayContains(window.location.host, result.blacklistSites.split(", ")))) {
```
Comment (simonfrey) [inline js/lazy_images.js]: Please merge the two `if` and add a comment what it does, so future developers know what the if does

File: js/instant_page.min.js [hunk 3]
```diff
@@ -1,18 +1,28 @@
...
+	  if (result.active == false) {
+		console.log("[faster pageload plugin] Inactive: Do not load instant.page script")
+		return
+	  }
+	  if (result.preloadViewport == true) {
+		document.body.dataset.instantIntensity = "viewport-all";
+	  }
+	  
+	  if (!(arrayContains(window.location.host, result.blacklistSites.split(", ")))) {
```
Comment (simonfrey) [inline js/instant_page.min.js]: Please use the [happy path scheme](https://en.wikipedia.org/wiki/Happy_path) as it gives less indentation and creates a better to read code.

```
if (arrayContains(window.location.host, result.blacklistSites.split(", "))) { 
  return
}
```

File: js/lazy_images.js (hunk 2 above)
Comment (simonfrey) [inline js/lazy_images.js]: Please use the [happy path scheme](https://en.wikipedia.org/wiki/Happy_path) as it gives less indentation and creates a better to read code.

```
if (!(result.imgLazyLoad == true) || arrayContains(window.location.host, result.blacklistSites.split(", "))) { 
  return
}
```

File: js/instant_page.min.js (hunk 3 above)
Comment (simonfrey) [reply js/instant_page.min.js]: I think `result.blacklistSites.split(", ")` explodes when `result.blacklistSites` is `undefined`. Did you check that?

**Review by simonfrey — CHANGES_REQUESTED** [review]
Thanks for the PR. Nice that you figured out how to program the feature!
Please check for the comments as I think they give the code a better readability.

And please do not forget to test for certain edge-cases for the blacklist. e.g. `""` `" "` `",,"` `"www.example.com,,"` `", , "`

**Comment by CactiChameleon9** [comment]
Yep, thanks those changes make sense

I will get back to you soon - and the happy path scheme seems good - I have never used that before.

I will probably start fresh with a few of the files, and apply the changes again so the commit changes don't look as strange (too much debugging :D)

**Comment by CactiChameleon9** [comment]
Yep, thanks those changes make sense

I will get back to you soon - and the happy path scheme seems good - I have never used that before.

I will probably start fresh with a few of the files, and apply the changes again so the commit changes don't look as strange

**Comment by simonfrey** [comment]
You can just add commits to the branch. That should work. If you want to cleanup your git history checkout git interactive rebase: https://www.youtube.com/watch?v=ElRzTuYln0M

File: js/instant_page.min.js (hunk 3 above)
Comment (CactiChameleon9) [reply js/instant_page.min.js]: How can I test is something explodes?

File: js/lazy_images.js (hunk 2 above)
Comment (CactiChameleon9) [reply js/lazy_images.js]: I instead used the happy path scheme - is that OK?

Comment (CactiChameleon9) [reply js/lazy_images.js]: Done

File: js/instant_page.min.js (hunk 3 above)
Comment (CactiChameleon9) [reply js/instant_page.min.js]: I used this to help with a blank:
  if (result.blacklistSites !== "" && result.blacklistSites.split(", ").includes(window.location.host)){
I think it should work because apparently && is a short-circuit meaning the second part should not run.

File: js/lazy_images.js (hunk 2 above)
Comment (CactiChameleon9) [reply js/lazy_images.js]: Ah... I did them in separate if statements - is that fine or do they need to be in one?

**Comment by CactiChameleon9** [comment]
OK... I should have done the changes, but testing is required. Also how would I test for certain edge-cases in the blacklist. If it just does not work because it is entered in wrong is that OK? Or am I looking for a crash?

File: js/instant_page.min.js (hunk 3 above)
Comment (simonfrey) [reply js/instant_page.min.js]: You cloud just write a short function testing certain strings, or you just manually test it by hand within in your activated plugin.

File: js/lazy_images.js (hunk 2 above)
Comment (simonfrey) [reply js/lazy_images.js]: That's fine. Just different coding styles. maybe two is even better understandable

**Comment by simonfrey** [comment]
No just check if it working or not. I don't think it could crash

File: js/instant_page.min.js (hunk 3 above)
Comment (CactiChameleon9) [reply js/instant_page.min.js]: OK - so just check that it is all OK from the end user within the addon

**Comment by CactiChameleon9** [comment]
I think it its all good now - just have a look.
Also blacklist seems fine, and only doesn't disable site entered in incorrectly with extra commas and spaces .

**Comment by simonfrey** [comment]
Sorry for the late response! Could you please add the functionality to also accept urls with the wrong amount of spaces?
E.g. you could [remove all spaces](https://stackoverflow.com/questions/5963182/how-to-remove-spaces-from-a-string-using-javascript) in the string and than the matching should work if you just split by the `,`

**Comment by CactiChameleon9** [comment]
No worries, we all have busy lives that can get in the way of hobbies. I will try and implement that and get back to you. Do you want me to change the note(on the options page), or leave it with a comma and space?

**Comment by simonfrey** [comment]
Leave it with the comma & space. Looks nicer.

**Comment by CactiChameleon9** [comment]
Done. I did the code in the options page save because doing extra js computing while loading a page could slow down the browser. I also added lowercasing of the URLs.

**Review by simonfrey — APPROVED** [review]
Thanks you for your work on this!

**Comment by CactiChameleon9** [comment]
No problem, glad I could help

**Comment by simonfrey** [comment]
Is deployed as v1.8 to the store: https://addons.mozilla.org/en-US/firefox/addon/faster-pageload/
